### ISSUE TITLE
1.4.0-beta.1 issues

### ISSUE THREAD

**Issue by samarulmeu** [Zettlr/Zettlr#290]
I again start by thanking you for your great and impressive work, but I think that you should know by now how much we appreciate what you do for us with this app. 

I must say that I am really impressed with the new beta. Most of the old issues are solved and moving files between folders is smooth now -- even between external storage devices. 

Now some issues I encountered (I may add some later here):

- the variables for the sidebar where changed so I had to tweak my `custom.css`, but this was not such a big deal.  
- when I select a tag from the tag list, let's say something like `#test`, after I press enter I get something like`#test undefined`
- the `Test your ID generation` button doesn't work 
- closing the app from the x window button let's Zettlr processes hanging in the background, `Ctrl+Q` closes all the processes. 
- there is an annoying refresh of the sidebar. 
- as a side question: will it be possible to have the app icon change color based on it's theme?

**Comment by nathanlesage** [comment]
> files between folders is smooth now -- even between external storage devices.

Ahh, perfect :) Would you mind looking through the many issues by you and see whether or not we can either close them altogether or merge them into one issue? This would help keep the issue list lean and clean while I don't lose track of all that's happening! :)

Now to your points:

* Yep, the DOM structure needed to change. On the one hand, all Custom CSS needs to be rewritten now, but I think this is a small price to pay for a now much better sidebar. Besides, many elements of the DOM are now similar, and therefore there is less room for friction or mistakes (the functional elements in both the tree view and the file list are now named and structured exactly the same which makes it easy to extract information — besides, I was finally able to rename "preview" to "file list", "directories" to "file tree" and "combiner" to "sidebar", which makes much more sense from a semantic point of view!)
- [ ] This issue was introduced when I enabled the automatic insertion of link titles, thanks for pointing it out, I'll fix it within the hour! ([Responsible line](https://github.com/Zettlr/Zettlr/blob/b1e8927db7c5b56ef6f8c1f264f226367dc4e839/source/renderer/zettlr-editor.js#L147))
- [ ] Small mistake by me, I simply forgot to rename some require-statements when I moved out the helpers into their respective utility-files. ([Responsible line](https://github.com/Zettlr/Zettlr/blob/b1e8927db7c5b56ef6f8c1f264f226367dc4e839/source/renderer/dialog/preferences.js#L154))
- [ ] I'll have a look into it. If I understand you correctly, closing the window by clicking the "X" does not quit the app, right?
* The refresh of the sidebar is necessary, because currently, the `vue-virtual-scroller` which acts as the replacement for `Clusterize.js` has some hiccups if the elements in the list change. Therefore I have to empty the element list for a short amount of time (~10ms) before re-filling it again. We all hope that the corresponding issue (https://github.com/Akryum/vue-virtual-scroller/issues/256) will be fixed soon …
* Concerning the application icon change: Would be really great, but while on Windows and Linux you can change the icon easily, on macOS one would need a workaround. I'll see to it, but it doesn't have any priority right now, as I just want to get 1.4 to work smoothly :)

**Comment by samarulmeu** [comment]
> Ahh, perfect :) Would you mind looking through the many issues by you and see whether or not we can either close them altogether or merge them into one issue? This would help keep the issue list lean and clean while I don't lose track of all that's happening! :)

Done! 

>  I'll have a look into it. If I understand you correctly, closing the window by clicking the "X" does not quit the app, right?

Yes. It is something I also mentioned in #263. The window is closed, but if you do `ps aux | grep zettlr`you find all the processes active. As opposed to `Ctrl+Q` which closes all processes. 

>  Concerning the application icon change: Would be really great, but while on Windows and Linux you can change the icon easily, on macOS one would need a workaround. I'll see to it, but it doesn't have any priority right now, as I just want to get 1.4 to work smoothly :)

It's not urgent by any means. Just a suggestion. 

But now I discovered a new **issue**: Writing Target indicator does not work. The color doesn't change to indicate the slice done.

**Comment by nathanlesage** [comment]
> But now I discovered a new issue: Writing Target indicator does not work. The color doesn't change to indicate the slice done.

Aaaaargh :D Okay, I'll see into it. This is the biggest pain in having a really huge software application to maintain — change something here, and by that break something somewhere else =D

**Comment by nathanlesage** [comment]
NB to the zombie-issue: I believe, this event is not correctly emitted:

https://github.com/Zettlr/Zettlr/blob/b1e8927db7c5b56ef6f8c1f264f226367dc4e839/source/main.js#L115-L126

**Comment by samarulmeu** [comment]
I see the logic. It will be interesting to see what is the behavior in Windows or other Linux distros (mine is Ubuntu based).

**Comment by nathanlesage** [comment]
Alright, at least on a macOS-system, the event is emitted as it should. P.S.: Just switched the GUI manually to the Linux-mode, and I definitely have to create different Linux window controls, they look embarrassing …

**Comment by samarulmeu** [comment]
Thank you! And I think that 1.4 will be a very nice and stable release with a lot of great feature.

**Comment by samarulmeu** [comment]
Although it is more a feature request than an issue, I thought I mention it here and maybe it will make it in the 1.4 release. 

So, will it be possible that when you try to move a file into a directory where it already exists one with the same name to have the option to overwrite?

**Comment by samarulmeu** [comment]
Will the overwriting option be hard to implement?

**Comment by nathanlesage** [comment]
Nope, it's just a little bit of messing around with these lines of code to not abort, but offer to override the file:

https://github.com/Zettlr/Zettlr/blob/ff831fcbaa9466a7bd6608a7932db9590eef80b2/source/main/commands/request-move.js#L55-L62

**Comment by samarulmeu** [comment]
Thank you!

**Comment by samarulmeu** [comment]
Can you please check the following errors? 

![error1](https://user-images.githubusercontent.com/39216570/63431626-ed914500-c427-11e9-8048-52bd9573e356.png)

![error2](https://user-images.githubusercontent.com/39216570/63431635-f2ee8f80-c427-11e9-9588-d36c616570b4.png)

I started to get them today. And Zettlr just refuses to open directories. The only way is to recover an older backup. Could it be related to `.ztr-directory` files?

**Comment by nathanlesage** [comment]
This seems most likely to be an error with these .ztr-directory files. Does the error go away after removing those files? (I thought that I had written a more or less good transforming function)

**Comment by samarulmeu** [comment]
The problem is that I don't know which files to blame. I have a lot of subdirectories in the main **opened** ones.

And also I found a strange behaviour with the `.ztr-directory` files: I am using rsync to backup my notes and it **always** sees this files as changed although in some subdirectories I didn't changed the sorting order.

**Comment by samarulmeu** [comment]
As a side note: will it be hard to implement only **one** `.ztr-directory` located in `.config`?

**Comment by nathanlesage** [comment]
It would be easy, but I can't do this, as it would undermine the whole cloud-sync-part that's the reason for having these files.

**Comment by samarulmeu** [comment]
Or maybe just one in the opened directory and not one for each subdirectory.

**Comment by nathanlesage** [comment]
Still, this would remove all settings in case you move a subdirectory. I mean, these files never caused any trouble, so I think just eliminating them would be too much ;)

**Comment by samarulmeu** [comment]
OK. But the mentioned errors still remain.

**Comment by nathanlesage** [comment]
After removing the questionable `.ztr-directory`-files?

**Comment by nathanlesage** [comment]
Still, after all I can't explain why the app would write malformed JSON… wait a moment. Could it be that some of the files became corrupted? Because the app itself is supposed to write clean JSON into these files, but the errors clearly state that the JSON was malformed o.O

**Comment by samarulmeu** [comment]
Tomorrow I'll look in all the files to see what is happening, although I don't understand why they get suddenly corrupted.

**Comment by nathanlesage** [comment]
me neither … this is something weird happening

**Comment by samarulmeu** [comment]
I think I managed to understand what is happening. 

In one of the open directories the `.ztr-directory` files looks like this:

```
{"sorting":"time-up","virtualDirectories":[],"settings":{"sorting":"name-up"}}}}
```

I have no idea from where the extra `}}` appeared. And it is strange that after removing **only** this file Zettlr opened the other directories too without complaining(I work with 4 opened directories, only in one the `.ztr-directory` looked like this). Before it displayed a message for every directory that it can not be opened.

**Comment by nathanlesage** [comment]
So it definitely sounds as if this file just decided to corrupt itself :/ But it's good that it hasn't happened with the other ones :)

**Comment by samarulmeu** [comment]
I don't know what happened. The thing is that the corrupted files was in a directory that is attached to a Nextcloud account (local folder syncing remotely). Let's suppose that something strange happened during sync (although the local files should have been fine), the problem was that Zettrl did refuse to open the other folders. My work place is something like 4 main folders: Notes (first ones Zettlr displays - the one connected to the Nextcloud account) and 3 more (let's say folder1, folder2, and folder3). Zettlr was stuck after trying to open Notes, refusing to open folder1 and the others, which I think wasn't suppose to happen. And the files in Notes were shown, that's way I didn't suspect that there was the problem there and I was looking in the wrong places. 

And also is this other thing that I don't understand. Why Zettlr rewrites every time it opens or closes (I'm not sure when) the `.ztr-directory` as `rsync` always considered them changed?

**Comment by nathanlesage** [comment]
That Zettlr refused to open the other folders is simply because I forgot a failsafe to fall back to default settings if the JSON is not safe to eval, which I need to address (cf. https://github.com/Zettlr/Zettlr/blob/35c63b03def434171c16f050193416baaf440006/source/main/zettlr-dir.js#L792), then Zettlr's parsing won't crash if there's a malformed file.

As to the writing … so rsync looks at the modification date, right? Then I'll need to check whether or not the settings have changed during runtime … good to know!

**Comment by samarulmeu** [comment]
> As to the writing … so rsync looks at the modification date, right? Then I'll need to check whether or not the settings have changed during runtime … good to know!

It seems that Zettlr changes all the `.ztr-directory` files when it starts and I don't understand why. When I close it the only affected files are the ones changed because I modified the sorting of the directory. 

My workaround now is to just exclude all `.ztr-directory` files from backup (rsync), but that contradicts the cloud-sync-part you've mentioned. But now I can see what I backup and not have to dig through a lot of screen lines...

**Comment by nathanlesage** [comment]
I just added the failsafe, so that corrupted files won't crash the initial directory tree loading.

**Comment by samarulmeu** [comment]
Hope it will solve the problem as for me is getting more and more a problem. Now it complains for other directories too. I started to delete all the `.ztr-directory` files, even in subfolders.

It is strange that there are no more complaints. I have no ideas what is **really** causing it: the rsync, the app, the Linux file system?

**Comment by nathanlesage** [comment]
I wish I knew the answer :/ On my system I've trained the conversion-algorithm from the old `.ztr-directory` to the new one, and encountered no errors afterwards … let's hope that this might be related to rsync, because then we would have an issue to grasp (not like the scrolling-error with `Clusterize.js` that we've had for MONTHS until I finally dropped it for the Vue.js-component …)

**Comment by samarulmeu** [comment]
Maybe the failsafe will solve the problem... 

I still think that a central file in an open directory would be a better option. This way Zettlr will remain the **clean** app that it is -- not interfering with the notes. Concerning the moving/renaming subfolders -- then the config will return to default until you change it in that particular subfolder. Only an idea.

**Comment by nathanlesage** [comment]
This is what I've already implemented: The `.ztr-directory`-file will only be present, if the directory is not default (e.g., another sorting than name ascending, or virtual directories). In any other case, the file should be removed during the update, so that we only have these files where necessary.

**Comment by nathanlesage** [comment]
Fixed the writing targets. So now I'll release a new beta and let's hope that the closing works now …

**Comment by samarulmeu** [comment]
Great. I was also thinking of a new beta, but I didn't dare to ask for it.

**Comment by nathanlesage** [comment]
:D Yeah, I think I'll postpone the remaining three issues I originally planned for 1.4 to 1.5 just to get out a new release — 1.3 is already getting old, I think.

**Comment by samarulmeu** [comment]
Yes, and 1.4 will be a nice start of autumn / September. I'll test it intensely in the beginning of the week.

**Comment by samarulmeu** [comment]
Closed as issues are solved by the new 1.4-0.beta.2
